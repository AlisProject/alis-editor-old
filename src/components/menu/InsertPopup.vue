<template>
  <div>
    <div
      class="insert-popup"
      :class="{ 'is-active': isActive, 'is-link': isClickedLink }"
      :style="{
        left: `${this.position.left}px`,
        top: `${this.position.top}px`
      }"
    >
      <button
        class="insert-popup__item"
        :class="[isBold ? 'insert-popup__bold_selected' : 'insert-popup__bold']"
        v-if="!isClickedLink"
        @click="execBold(activeRoot)"
      ></button>
      <button
        class="insert-popup__item"
        :class="[isItalic ? 'insert-popup__italic_selected' : 'insert-popup__italic']"
        v-if="!isClickedLink"
        @click="execItalic(activeRoot)"
      ></button>
      <button
        class="insert-popup__item"
        :class="[isQuote ? 'insert-popup__quote_selected' : 'insert-popup__quote']"
        v-if="!isClickedLink"
        @click="execQuote(activeRoot)"
      ></button>
      <button
        class="insert-popup__item"
        :class="[isHeading ? 'insert-popup__heading_selected' : 'insert-popup__heading']"
        v-if="!isClickedLink"
        @click="execHeading(activeRoot)"
      ></button>
      <button
        class="insert-popup__item"
        :class="[isSubHeading ? 'insert-popup__subheading_selected' : 'insert-popup__subheading']"
        v-if="!isClickedLink"
        @click="execSubHeading(activeRoot)"
      ></button>
      <button
        class="insert-popup__item"
        :class="[isAnchor ? 'insert-popup__textlink_selected' : 'insert-popup__textlink']"
        v-if="!isClickedLink"
        @click="execLink"
      ></button>
      <input class="insert-link" type="text" v-if="isClickedLink" :value="value" @input="handleInput" />
      <button class="insert-popup__item handle-url" style="cursor: pointer; color: #232538;" v-if="isClickedLink" @click="setUrl">✓</button>
      <button class="insert-popup__item handle-url" style="cursor: pointer; color: #232538;" v-if="isClickedLink" @click="deletePopup">×</button>
      <!--
        <button class="insert-popup__item" @click="execSubHeading">
          <InsertPopupIcon :src="require('../../assets/icon-h3.svg.js')" />
        </button>
      -->
    </div>
    <div
      class="link-popup"
      :class="{ 'is-active': isPopupActive }"
      :style="{
        left: `${this.hoverPosition.left}px`,
        top: `${this.hoverPosition.top}px`,
        cursor: 'pointer'
      }"
      @click="openInsertLink"
    >
      {{ this.url }}
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import InsertPopupIcon from './InsertPopupIcon.vue'
import { BlockType } from '../../types/Blocks'
import * as SvgIcon from '../vector/SvgIcon'
import { isContentEditableBlock } from '../../utils/createBlock'
import * as commandExecuter from '../../utils/commandExecuter'
import { getEndNodeByElement } from '../../utils/browserSelection'
import { findTreeContentById } from '../../utils/treeUtil'

export default Vue.extend({
  props: {
    activeRoot: Object,
    popUpPosition: Object,
    hoverPopupPosition: Object,
    isHover: Boolean,
    linked_url: String,
    targetAnchorNode: Object,
    store: Object
  },
  data(): any {
    return {
      isOpen: false,
      isActive: false,
      el: null,
      position: {
        left: 0,
        top: 0
      },
      hoverPosition: {
        left: 0.0,
        top: 0.0
      },
      isLink: false,
      isClickedLink: false,
      value: '',
      range: null,
      isPopupActive: this.isHover as boolean,
      url: this.linked_url as string,
      targetAnchor: null,
      isBold: false,
      isAnchor: false,
      isItalic: false,
      isQuote: false,
      isSubHeading: false,
      isHeading: false
    }
  },
  components: {
    InsertPopupIcon
  },
  computed: {
    SvgIcon() {
      return SvgIcon
    }
  },
  mounted() {
    document.addEventListener('mousedown', (event: any) => {
      if (this.$el.contains(event.target)) {
        return
      }
      this.isActive = false
      this.isClickedLink = false
      this.value = ''
    })
    document.addEventListener('selectionchange', this.onSelectionChange)
  },
  beforeDestroy() {
    document.removeEventListener('selectionchange', this.onSelectionChange)
  },
  watch: {
    activeRoot(aR) {
      this.el = document.querySelector(`[data-block-id="${aR.id}"]`)
      this.position.left = (document.querySelector('#ALISEditor') as any).offsetWidth * 0.5 - 110
      this.position.top = this.el.getBoundingClientRect().top - 50 + window.pageYOffset
      if (isContentEditableBlock(aR)) {
        this.isActive = false
      } else if (!['Image', 'Embed', 'Rule'].includes(aR.type)) {
        this.isActive = true
      }
    },
    'popUpPosition.left'() {
      this.position.left = this.popUpPosition.left
    },
    'popUpPosition.top'() {
      this.position.top = this.popUpPosition.top
    },
    'hoverPopupPosition.left'() {
      this.hoverPosition.left = this.hoverPopupPosition.left
    },
    'hoverPopupPosition.top'() {
      this.hoverPosition.top = this.hoverPopupPosition.top
    },
    isHover() {
      this.isPopupActive = this.isHover
    },
    linked_url() {
      this.url = this.linked_url
    },
    targetAnchorNode() {
      this.targetAnchor = this.targetAnchorNode
    }
  },
  methods: {
    handleInput(event: Event) {
      this.isActive = true
      this.value = (event.target as any).value
      if (this.targetAnchor !== null) {
        this.targetAnchor.href = this.value

      }
    },
    onSelectionChange(event: Event) {
      if (!this.activeRoot) {
        return
      }
      if (!isContentEditableBlock(this.activeRoot)) {
        return
      }
      const selection = window.getSelection()
      if (selection.isCollapsed && this.isClickedLink === false) {
        this.isActive = false
        return
      }
      // ポップアップの状態を初期化し、現在セレクションを始めたノードが何のタグに囲まれているノードかを判定する
      ;(async () => {
        await this.initiatePopupState()
        this.popupStateChange(selection.anchorNode)
      })()

      // Coordinate calculation of popup display
      const alisEditor = document.getElementById('ALISEditor')
      const alisEditorRect = (alisEditor as any).getBoundingClientRect()
      const range = selection.getRangeAt(0).cloneRange()
      if ((range as any).getClientRects().length !== 0) {
        const rect = (range as any).getClientRects()[0]
        const absoluteRectLeft = (rect.left + rect.right) / 2 - alisEditorRect.left - 105
        this.position.left = absoluteRectLeft
        const absoluteRectTop = rect.top - alisEditorRect.top - 75
        this.position.top = absoluteRectTop
      }

      this.isActive = true
    },
    ...commandExecuter,
    execLink() {
      if (this.isAnchor) {
        return document.execCommand('unlink')
      }
      this.isClickedLink = true
      const selection = window.getSelection()
      this.range = selection.getRangeAt(0)
      this.position.left = this.position.left - 125
    },
    setUrl() {
      ;(async () => {
        const selection = window.getSelection()
        selection.removeAllRanges()
        if (this.range !== null) {
          selection.addRange(this.range)
          document.execCommand('unlink')
          document.execCommand('createLink', true, this.value)
          await this.$nextTick()
          const aEl = selection.focusNode.parentElement as any
          aEl.addEventListener('mouseover', this.hoverEvent)
          aEl.addEventListener('mouseleave', this.deleteHover)
        }
        this.value = ''
        this.isActive = false
        this.isClickedLink = false
        this.targetAnchor = null
        this.$emit('deleteTargetAnchorNode')
        this.$emit('update', findTreeContentById(this.activeRoot.id, this.store.state.blocks))
        this.range = null
        selection.removeAllRanges()
      })()
    },
    deletePopup() {
      this.value = ''
      this.isActive = false
      this.isClickedLink = false
      this.range = null
      this.targetAnchor = null
      this.$emit('deleteTargetAnchorNode')
    },
    hoverEvent(event: any) {
      this.isPopupActive = true
      this.targetAnchor = event.target
      this.url = event.target.href
      // elementの座標を取得し、windowの絶対座標をマウスオーバー時に出現するポップアップの座標としてセットする
      const rect = event.srcElement.getBoundingClientRect()
      const alisEditor = document.getElementById('ALISEditor')
      const alisEditorRect = (alisEditor as any).getBoundingClientRect()
      const absoluteUnderPopupRectLeft = rect.left + rect.width / 2 - alisEditorRect.left - 175
      const absoluteUnderPopupRectTop = rect.top - alisEditorRect.top + 30
      this.hoverPosition.left = absoluteUnderPopupRectLeft
      this.hoverPosition.top = absoluteUnderPopupRectTop
    },
    openInsertLink(event: any) {
      const rect = event.srcElement.getBoundingClientRect()
      const alisEditor = document.getElementById('ALISEditor')
      const alisEditorRect = (alisEditor as any).getBoundingClientRect()
      const absoluteUpperPopupRectLeft = rect.left + rect.width / 2 - alisEditorRect.left - 265
      const absoluteUpperPopupRectTop = rect.top - alisEditorRect.top - 90
      this.position.left = absoluteUpperPopupRectLeft
      this.position.top = absoluteUpperPopupRectTop
      this.isPopupActive = false
      this.isClickedLink = true
      this.isActive = true
      this.value = this.url
    },
    deleteHover() {
      setTimeout(() => {
        this.isPopupActive = false
      }, 1000)
    },
    popupStateChange(targetNode: any) {
      const requireRecursivenodeNames = ['#text', 'A', 'H2', 'H3', 'I', 'B']
      const processTerminateNodes = ['BLOCKQUOTE', 'P']
      if (requireRecursivenodeNames.includes(targetNode.nodeName)) {
        const nodeName = targetNode.nodeName
        switch (nodeName) {
          case 'A':
            this.isAnchor = true
            break
          case 'H2':
            this.isHeading = true
            break
          case 'H3':
            this.isSubHeading = true
            break
          case 'I':
            this.isItalic = true
            break
          case 'B':
            this.isBold = true
            break
        }
        const parentNode = targetNode.parentNode
        return (this as any).popupStateChange(parentNode)
      } else if (processTerminateNodes.includes(targetNode.nodeName)) {
        if (targetNode.nodeName === 'BLOCKQUOTE') {
          this.isQuote = true
        }
        return
      }
    },
    initiatePopupState() {
      this.isBold = false
      this.isAnchor =  false
      this.isItalic = false
      this.isQuote = false
      this.isSubHeading = false
      this.isHeading = false
    }
  }
})
</script>

<style scoped>
.insert-popup {
  margin: 0;
  padding: 0;
  color: #fff;
  display: flex;
  list-style-type: none;
  font-size: 1.4rem;
  background-color: #fff;
  width: 300px;
  position: absolute;
  z-index: 200;
  opacity: 0;
  pointer-events: none;
  border-radius: 5px;
  box-shadow: 0 0 10px gray;
}

.insert-popup.is-active {
  pointer-events: initial;
  animation: anim1 0.2s cubic-bezier(0.175, 0.885, 0.32, 1.275) forwards;
}

.insert-popup.is-active.is-link {
  pointer-events: initial;
  animation: anim1 0.2s cubic-bezier(0.175, 0.885, 0.32, 1.275) forwards;
  width: 500px;
}

.link-popup {
  margin: 0;
  padding: 10px;
  color: #fff;
  text-align: center;
  display: flex;
  list-style-type: none;
  font-size: 1.4rem;
  background: #232538;
  width: 300px;
  border-radius: 5px;
  position: absolute;
  z-index: 200;
  opacity: 0;
  pointer-events: none;
  word-break: break-all;
}

.link-popup.is-active {
  pointer-events: initial;
  animation: anim1 0.2s cubic-bezier(0.175, 0.885, 0.32, 1.275) forwards;
}

@keyframes anim1 {
  0% {
    transform: translateY(10px);
    opacity: 0;
  }
  100% {
    transform: translateY(0px);
    opacity: 1;
  }
}

/*.insert-popup::after {*/
/*position: absolute;*/
/*left: calc(50% - 10px);*/
/*bottom: -10px;*/
/*width: 20px;*/
/*height: 20px;*/
/*background: #41446a;*/
/*transform: rotate(45deg);*/
/*content: '';*/
/*display: block;*/
/*}*/

.insert-popup .insert-popup__item {
  /*margin-left: 8px;*/
  width: 50px;
  height: 50px;
  display: flex;
  align-items: center;
  justify-content: center;
  color: #fff;
  font-weight: bold;
  padding-left: 1px;
  padding-bottom: 1px;
  transition: all 0.05s linear;
  transform: rotate(0deg);
  font-size: 25px;
  font-weight: normal;
  font-family: 'Yu Gothic', YuGothic;
  appearance: none;
  -webkit-appearance: none;
  border: none;
  outline: none;
  cursor: pointer;
}

.insert-popup__bold {
  background: #fff url('../../../assets/images/pc/editor/icon_bold.png') no-repeat;
  background-size: 32px;
  background-position: 8px 10px;
  border-radius: 5px 0px 0px 5px;
}

.insert-popup__bold_selected {
  background: #232538 url('../../../assets/images/pc/editor/icon_bold_selected.png') no-repeat;
  background-size: 32px;
  background-position: 8px 10px;
  border-radius: 5px;
  border-radius: 5px 0px 0px 5px;
}

.insert-popup__italic {
  background: #fff url('../../../assets/images/pc/editor/icon_italic.png') no-repeat;
  background-size: 32px;
  background-position: 8px 10px;
}

.insert-popup__italic_selected {
  background: #232538 url('../../../assets/images/pc/editor/icon_italic_selected.png') no-repeat;
  background-size: 32px;
  background-position: 8px 10px;
}

.insert-popup__quote {
  background: #fff url('../../../assets/images/pc/editor/icon_quote.png') no-repeat;
  background-size: 32px;
  background-position: 8px 10px;
}

.insert-popup__quote_selected {
  background: #232538 url('../../../assets/images/pc/editor/icon_quote_selected.png') no-repeat;
  background-size: 32px;
  background-position: 8px 10px;
}

.insert-popup__heading {
  background: #fff url('../../../assets/images/pc/editor/icon_index.png') no-repeat;
  background-size: 32px;
  background-position: 8px 10px;
  cursor: pointer;
}

.insert-popup__heading_selected {
  background: #232538 url('../../../assets/images/pc/editor/icon_index_selected.png') no-repeat;
  background-size: 32px;
  background-position: 8px 10px;
  cursor: pointer;
}

.insert-popup__subheading {
  background: #fff url('../../../assets/images/pc/editor/icon_text.png') no-repeat;
  background-size: 32px;
  background-position: 8px 10px;
}

.insert-popup__subheading_selected {
  background: #232538 url('../../../assets/images/pc/editor/icon_text_selected.png') no-repeat;
  background-size: 32px;
  background-position: 8px 10px;
}

.insert-popup__textlink {
  background: #fff url('../../../assets/images/pc/editor/icon_textlink.png') no-repeat;
  background-size: 32px;
  background-position: 8px 10px;
  border-radius: 0px 5px 5px 0px;
}

.insert-popup__textlink_selected {
  background: #232538 url('../../../assets/images/pc/editor/icon_textlink_selected.png') no-repeat;
  background-size: 32px;
  background-position: 8px 10px;
  border-radius: 0px 5px 5px 0px;
}

.insert-popup .insert-link {
  margin-left: 8px;
  border-radius: 5px;
  width: 400px;
  height: 50px;
  display: flex;
  align-items: center;
  justify-content: center;
  color: #232538;
  font-weight: bold;
  padding-left: 1px;
  padding-bottom: 1px;
  transition: all 0.05s linear;
  transform: rotate(0deg);
  font-size: 18px;
  font-weight: normal;
  font-family: 'Yu Gothic', YuGothic;
  appearance: none;
  -webkit-appearance: none;
  background: transparent;
  border: none;
  outline: none;
}

input {
  display: none;
}
</style>
