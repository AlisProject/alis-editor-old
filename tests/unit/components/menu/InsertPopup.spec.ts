import { mount } from '@vue/test-utils'
import InsertPopup from '@/components/menu/InsertPopup.vue'

describe('InsertPopup.vue', () => {
  test('renders correctly', () => {
    const wrapper = mount(InsertPopup)
    expect(wrapper.element).toMatchSnapshot()
  })
})
