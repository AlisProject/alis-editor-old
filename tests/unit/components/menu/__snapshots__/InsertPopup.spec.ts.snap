// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InsertPopup.vue renders correctly 1`] = `
<div>
  <div
    class="insert-popup"
    style="left: 0px; top: 0px;"
  >
    <button
      class="insert-popup__item"
    >
      <div
        class="editor-insert-popup__icon"
      >
        <div
          class="editor-insert-popup__iconContent"
        >
          <!--?xml version="1.0" encoding="UTF-8"?-->
          

          <svg
            height="17px"
            version="1.1"
            viewBox="0 0 12 17"
            width="12px"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            
    
            <!-- Generator: Sketch 52 (66869) - http://www.bohemiancoding.com/sketch -->
            
    
            <title>
              B
            </title>
            
    
            <desc>
              Created with Sketch.
            </desc>
            
    
            <g
              fill="none"
              fill-rule="evenodd"
              id="SP"
              stroke="none"
              stroke-width="1"
            >
              
        
              <g
                class="fill-target"
                id="ストーリー_マイストーリー_入力中_テキスト選択中"
                transform="translate(-34.000000, -14.000000)"
              >
                
            
                <g
                  id="Group-14"
                >
                  
                
                  <g
                    id="Group-55"
                    transform="translate(20.000000, 0.000000)"
                  >
                    
                    
                    <g
                      id="Group-4"
                    >
                      
                        
                      <g
                        id="icon_bold"
                      >
                        
                            
                        <g
                          id="Group-52"
                        >
                          
                                
                          <path
                            d="M14,14 L19.368,14 C23.262,14 24.758,15.452 24.758,17.982 C24.758,19.896 23.394,21.238 21.238,21.546 L21.238,21.722 C23.812,22.052 25.264,23.504 25.264,25.682 C25.264,28.52 23.196,30.214 19.676,30.214 L14,30.214 L14,14 Z M15.87,15.54 L15.87,20.908 L18.796,20.908 C21.436,20.908 22.822,19.83 22.822,18.114 C22.822,16.376 21.92,15.54 19.126,15.54 L15.87,15.54 Z M15.87,22.448 L15.87,28.674 L19.434,28.674 C22.074,28.674 23.284,27.574 23.284,25.594 C23.284,23.658 21.766,22.448 18.928,22.448 L15.87,22.448 Z"
                            id="B"
                          />
                          
                            
                        </g>
                        
                        
                      </g>
                      
                    
                    </g>
                    
                
                  </g>
                  
            
                </g>
                
        
              </g>
              
    
            </g>
            

          </svg>
        </div>
      </div>
    </button>
     
    <button
      class="insert-popup__item"
    >
      <div
        class="editor-insert-popup__icon"
      >
        <div
          class="editor-insert-popup__iconContent"
        >
          <!--?xml version="1.0" encoding="UTF-8"?-->
          

          <svg
            height="18px"
            version="1.1"
            viewBox="0 0 8 18"
            width="8px"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            
    
            <!-- Generator: Sketch 52 (66869) - http://www.bohemiancoding.com/sketch -->
            
    
            <title>
              i
            </title>
            
    
            <desc>
              Created with Sketch.
            </desc>
            
    
            <g
              fill="none"
              fill-rule="evenodd"
              id="SP"
              stroke="none"
              stroke-width="1"
            >
              
        
              <g
                class="fill-target"
                id="ストーリー_マイストーリー_入力中_テキスト選択中"
                transform="translate(-76.000000, -12.000000)"
              >
                
            
                <g
                  id="Group-14"
                >
                  
                
                  <g
                    id="Group-55"
                    transform="translate(20.000000, 0.000000)"
                  >
                    
                    
                    <g
                      id="Group-4"
                    >
                      
                        
                      <g
                        id="Group-54"
                        transform="translate(40.000000, 0.000000)"
                      >
                        
                            
                        <g
                          id="icon_italic"
                          transform="translate(4.000000, 6.000000)"
                        >
                          
                                
                          <g
                            id="Group-2"
                          >
                            
                                    
                            <path
                              d="M16.984,11.304 L15.016,11.304 L15.016,23.568 L16.984,23.568 L16.984,11.304 Z M17.2,6.576 L14.8,6.576 L14.8,8.952 L17.2,8.952 L17.2,6.576 Z"
                              id="i"
                              transform="translate(16.000000, 15.072000) scale(-1, 1) rotate(-15.000000) translate(-16.000000, -15.072000) "
                            />
                            
                                
                          </g>
                          
                            
                        </g>
                        
                        
                      </g>
                      
                    
                    </g>
                    
                
                  </g>
                  
            
                </g>
                
        
              </g>
              
    
            </g>
            

          </svg>
        </div>
      </div>
    </button>
     
    <button
      class="insert-popup__item"
    >
      <div
        class="editor-insert-popup__icon"
      >
        <div
          class="editor-insert-popup__iconContent"
        >
          <!--?xml version="1.0" encoding="UTF-8"?-->
          

          <svg
            height="12px"
            version="1.1"
            viewBox="0 0 18 12"
            width="18px"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            
    
            <!-- Generator: Sketch 51.2 (57519) - http://www.bohemiancoding.com/sketch -->
            
    
            <title>
              ‘‘
            </title>
            
    
            <desc>
              Created with Sketch.
            </desc>
            
    
            <defs />
            
    
            <g
              fill="none"
              fill-rule="evenodd"
              id="SP"
              stroke="none"
              stroke-width="1"
            >
              
        
              <g
                class="fill-target"
                id="ストーリー_マイストーリー_入力中_テキスト未選択"
                transform="translate(-115.000000, -16.000000)"
              >
                
            
                <g
                  id="Group-14"
                >
                  
                
                  <g
                    id="icon_quote"
                    transform="translate(104.000000, 0.000000)"
                  >
                    
                    
                    <g
                      id="Group-2"
                    >
                      
                        
                      <g
                        id="Group-11"
                      >
                        
                            
                        <path
                          d="M15.944,16 L13.736,16 L11,22.528 L11,27.568 L16.232,27.568 L16.232,22.48 L13.448,22.48 L15.944,16 Z M28.376,16 L26.168,16 L23.432,22.528 L23.432,27.568 L28.664,27.568 L28.664,22.48 L25.88,22.48 L28.376,16 Z"
                          id="‘‘"
                        />
                        
                        
                      </g>
                      
                    
                    </g>
                    
                
                  </g>
                  
            
                </g>
                
        
              </g>
              
    
            </g>
            

          </svg>
        </div>
      </div>
    </button>
     
    <button
      class="insert-popup__item"
    >
      <div
        class="editor-insert-popup__icon"
      >
        <div
          class="editor-insert-popup__iconContent"
        >
          <!--?xml version="1.0" encoding="UTF-8"?-->
          

          <svg
            height="18px"
            version="1.1"
            viewBox="0 0 14 18"
            width="14px"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            
    
            <!-- Generator: Sketch 51.2 (57519) - http://www.bohemiancoding.com/sketch -->
            
    
            <title>
              T
            </title>
            
    
            <desc>
              Created with Sketch.
            </desc>
            
    
            <defs />
            
    
            <g
              fill="none"
              fill-rule="evenodd"
              id="SP"
              stroke="none"
              stroke-width="1"
            >
              
        
              <g
                class="fill-target"
                id="ストーリー_マイストーリー_入力中_テキスト未選択"
                transform="translate(-35.000000, -13.000000)"
              >
                
            
                <g
                  id="Group-14"
                >
                  
                
                  <g
                    id="icon_index"
                    transform="translate(20.000000, 0.000000)"
                  >
                    
                    
                    <g
                      id="Group-2"
                    >
                      
                        
                      <g
                        id="Group-6"
                      >
                        
                            
                        <polygon
                          id="T"
                          points="28.152 13 15 13 15 14.776 20.496 14.776 20.496 30.688 22.608 30.688 22.608 14.776 28.152 14.776"
                        />
                        
                        
                      </g>
                      
                    
                    </g>
                    
                
                  </g>
                  
            
                </g>
                
        
              </g>
              
    
            </g>
            

          </svg>
        </div>
      </div>
    </button>
     
    <button
      class="insert-popup__item"
    >
      <div
        class="editor-insert-popup__icon"
      >
        <div
          class="editor-insert-popup__iconContent"
        >
          <!--?xml version="1.0" encoding="UTF-8"?-->
          

          <svg
            height="13px"
            version="1.1"
            viewBox="0 0 9 13"
            width="9px"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            
    
            <!-- Generator: Sketch 51.2 (57519) - http://www.bohemiancoding.com/sketch -->
            
    
            <title>
              icon_text
            </title>
            
    
            <desc>
              Created with Sketch.
            </desc>
            
    
            <defs />
            
    
            <g
              fill="none"
              fill-rule="evenodd"
              id="SP"
              stroke="none"
              stroke-width="1"
            >
              
        
              <g
                id="ストーリー_マイストーリー_入力中_テキスト未選択"
                transform="translate(-80.000000, -16.000000)"
              >
                
            
                <g
                  id="Group-14"
                >
                  
                
                  <g
                    id="Group-7"
                    transform="translate(64.000000, 0.000000)"
                  >
                    
                    
                    <g
                      id="icon_text"
                      transform="translate(4.000000, 6.000000)"
                    >
                      
                        
                      <g
                        id="Group-2"
                      >
                        
                            
                        <rect
                          height="32"
                          id="icon"
                          width="32"
                          x="0"
                          y="0"
                        />
                        
                            
                        <polygon
                          class="fill-target"
                          id="T"
                          points="20.884 10.92 12.116 10.92 12.116 12.104 15.78 12.104 15.78 22.712 17.188 22.712 17.188 12.104 20.884 12.104"
                        />
                        
                        
                      </g>
                      
                    
                    </g>
                    
                
                  </g>
                  
            
                </g>
                
        
              </g>
              
    
            </g>
            

          </svg>
        </div>
      </div>
    </button>
     
    <button
      class="insert-popup__item"
    >
      <div
        class="editor-insert-popup__icon"
      >
        <div
          class="editor-insert-popup__iconContent"
        >
          <!--?xml version="1.0" encoding="UTF-8"?-->
          

          <svg
            height="18px"
            version="1.1"
            viewBox="0 0 18 18"
            width="18px"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            
    
            <!-- Generator: Sketch 52 (66869) - http://www.bohemiancoding.com/sketch -->
            
    
            <title>
              icon_link
            </title>
            
    
            <desc>
              Created with Sketch.
            </desc>
            
    
            <g
              fill="none"
              fill-rule="evenodd"
              id="SP"
              stroke="none"
              stroke-width="1"
            >
              
        
              <g
                class="fill-target"
                id="ストーリー_マイストーリー_入力中_テキスト選択中"
                transform="translate(-111.000000, -13.000000)"
              >
                
            
                <g
                  id="Group-14"
                >
                  
                
                  <g
                    id="Group-55"
                    transform="translate(20.000000, 0.000000)"
                  >
                    
                    
                    <g
                      id="icon_textlink"
                      transform="translate(80.000000, 0.000000)"
                    >
                      
                        
                      <g
                        id="Group-2"
                      >
                        
                            
                        <g
                          id="Group-53"
                        >
                          
                                
                          <g
                            id="icon_link"
                            transform="translate(11.000000, 13.000000)"
                          >
                            
                                    
                            <path
                              d="M10.757,6.198 L6.1135,10.5495 C5.811,10.8325 5.7955,11.307 6.079,11.6095 C6.2265,11.767 6.4265,11.8465 6.6265,11.8465 C6.81,11.8465 6.9945,11.7795 7.139,11.644 L11.7825,7.2925 C12.085,7.0095 12.1005,6.5345 11.817,6.2325 C11.534,5.93 11.0595,5.915 10.757,6.198"
                              id="Fill-186"
                            />
                            
                                    
                            <path
                              d="M16.554,1.6115 C14.702,-0.2175 11.609,-0.1905 9.649,1.681 L7.0755,4.2365 C6.782,4.5285 6.7805,5.003 7.072,5.297 C7.3645,5.5915 7.8385,5.5925 8.133,5.301 L10.6955,2.756 C12.064,1.449 14.219,1.4145 15.5,2.6785 C16.088,3.259 16.406,4.067 16.3955,4.9535 C16.3845,5.8505 16.0295,6.694 15.421,7.266 L12.5865,9.9315 C12.285,10.2155 12.2705,10.69 12.554,10.9915 C12.7015,11.149 12.901,11.228 13.1005,11.228 C13.2845,11.228 13.469,11.1605 13.6145,11.0245 L16.4485,8.359 C17.3525,7.5085 17.8795,6.2735 17.895,4.971 C17.9105,3.6925 17.4215,2.468 16.554,1.6115"
                              id="Fill-187"
                            />
                            
                                    
                            <path
                              d="M9.849,12.5485 L7.286,15.0935 C5.917,16.399 3.761,16.4345 2.481,15.1705 C1.8935,14.5905 1.5755,13.7825 1.5865,12.896 C1.597,11.9985 1.952,11.156 2.5605,10.583 L5.395,7.9175 C5.6965,7.634 5.711,7.159 5.4275,6.8575 C5.1435,6.555 4.6695,6.541 4.367,6.8245 L1.5325,9.4905 C0.629,10.341 0.102,11.5755 0.0865,12.878 C0.0715,14.157 0.56,15.382 1.4275,16.238 C2.3415,17.14 3.557,17.5905 4.7905,17.5905 C6.0555,17.5905 7.3395,17.1155 8.3325,16.168 L10.906,13.613 C11.1995,13.321 11.2015,12.846 10.9095,12.552 C10.6175,12.258 10.143,12.257 9.849,12.5485"
                              id="Fill-188"
                            />
                            
                                
                          </g>
                          
                            
                        </g>
                        
                        
                      </g>
                      
                    
                    </g>
                    
                
                  </g>
                  
            
                </g>
                
        
              </g>
              
    
            </g>
            

          </svg>
        </div>
      </div>
    </button>
     
    <!---->
     
    <!---->
     
    <!---->
  </div>
   
  <div
    class="link-popup"
    style="left: 0px; top: 0px; cursor: pointer;"
  >
    
    
  
  </div>
</div>
`;
